mythril/laser/ethereum/instructions.py

```diff
@@ -76,11 +79,32 @@ def increment_states_pc(self, states: List[GlobalState]) -> List[GlobalState]:
                 state.mstate.pc += 1
         return states
 
+    @staticmethod
+    def check_gas_usage_limit(global_state: GlobalState):
+        global_state.mstate.check_gas()
+        if (
+            global_state.mstate.min_gas_used
+            >= global_state.current_transaction.gas_limit
+        ):
+            raise OutOfGasException()
+
+    def accumulate_gas(self, global_state: GlobalState):
+        if not self.enable_gas:
+            return global_state
+        opcode = global_state.instruction["opcode"]
+        min_gas, max_gas = OPCODE_GAS[opcode]
+        global_state.mstate.min_gas_used += min_gas
+        global_state.mstate.max_gas_used += max_gas
+        return global_state
+
     def __call__(self, func: Callable) -> Callable:
         def wrapper(
             func_obj: "Instruction", global_state: GlobalState
         ) -> List[GlobalState]:
             new_global_states = self.call_on_state_copy(func, func_obj, global_state)
+            new_global_states = [
+                self.accumulate_gas(state) for state in new_global_states
```

Comment:
Do you pay gas before or after the instruction execution?

Reply:
If you look at the pyethereum VM code, it seems like the gas is checked after the elements have been popped from the stack and before any state changes are made, e.g. [here](https://github.com/ethereum/pyethereum/blob/develop/ethereum/vm.py#L763). That shouldn't really affect us as we discard the global states anyway. At points where expensive memory operations can trigger an overflow, there is either an explicit gas check in the mstate's `mem_extend`, or in the state changing function.